contracts/iden3/LightweightState.sol

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.16;
+
+import {MerkleProof} from "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+
+import {UUPSSignableUpgradeable} from "@rarimo/evm-bridge-contracts/bridge/proxy/UUPSSignableUpgradeable.sol";
+import {Signers} from "@rarimo/evm-bridge-contracts/utils/Signers.sol";
+
+import {ILightweightState} from "../interfaces/ILightweightState.sol";
+
+/**
+ * @dev This contract is a copy of the LightweightState contract from Rarimo [identity-contracts repository](https://github.com/rarimo/identity-contracts/tree/aeb929ccc3fa8ab508fd7576f9fa853a081e5010).
+ */
+contract LightweightState is ILightweightState, UUPSSignableUpgradeable, Signers {
+    address public override sourceStateContract;
+    string public override sourceChainName;
+
+    bytes32 public override identitiesStatesRoot;
+
+    uint256 internal _currentGistRoot;
+
+    // gist root => GistRootData
+    mapping(uint256 => GistRootData) internal _gistsRootData;
+
+    // identities states root => identities states root data
+    mapping(bytes32 => IdentitiesStatesRootData) internal _identitiesStatesRootsData;
+
+    function __LightweightState_init(
+        address signer_,
+        address sourceStateContract_,
+        string calldata sourceChainName_,
+        string calldata chainName_
+    ) external initializer {
+        __Signers_init(signer_, chainName_);
+
+        sourceStateContract = sourceStateContract_;
+        sourceChainName = sourceChainName_;
+    }
+
+    function changeSigner(
+        bytes calldata newSignerPubKey_,
+        bytes calldata signature_
+    ) external override {
+        _checkSignature(keccak256(newSignerPubKey_), signature_);
+
+        signer = _convertPubKeyToAddress(newSignerPubKey_);
+    }
+
+    function changeSourceStateContract(
+        address newSourceStateContract_,
+        bytes calldata signature_
+    ) external override {
+        require(newSourceStateContract_ != address(0), "LightweightState: Zero address");
+
+        validateChangeAddressSignature(
+            uint8(MethodId.ChangeSourceStateContract),
+            address(this),
+            newSourceStateContract_,
+            signature_
+        );
+
+        sourceStateContract = newSourceStateContract_;
+    }
+
+    function signedTransitState(
+        bytes32 newIdentitiesStatesRoot_,
+        GistRootData calldata gistData_,
+        bytes calldata proof_
+    ) external override {
+        _checkMerkleSignature(_getSignHash(gistData_, newIdentitiesStatesRoot_), proof_);
+
+        require(
+            !isIdentitiesStatesRootExists(newIdentitiesStatesRoot_),
+            "LightweightState: Identities states root already exists"
+        );
+        require(
```

Comment:
This may revert on identical gist roots

Reply:
The method which calls the `signedTransitState` method does validation